### ISSUE TITLE
Setup a system for sharing the metadb with third parties

### ISSUE THREAD

**Issue by hellais** [ooni/sysadmin#272]
Currently the way by which we share the metadb (see "Metadata as a service" section of https://ooni.torproject.org/post/mining-ooni-data/), which is the postgresql database containing the metrics extracted from raw OONI measurements, is that we provide a static dump of the relevant database tables.

This is quite limited especially if the user of this database dump is interested in having fresh data.

We should provide a way for people interested in consuming OONI data to receive incremental updates to the postgresql database.

We are currently running postgresql 9.6, which has the following replication solutions: https://www.postgresql.org/docs/9.6/different-replication-solutions.html.

The challenge with the replication solutions present in 9.6 is that, by default it will replicate all the tables including ones which may contain secrets we don't want to publish.

If we do go for the approach of using a WAL log, we may even want to publish this log on s3 as this would also help us recover from a disaster.

Since postgresql 10 there is a feature called logical replication, which supports replicating on a per table basis: https://www.postgresql.org/docs/10/logical-replication.html.

The open questions to address for this are:

* [x] If we go the WAL log route, how do we sanitise/remove user password hashes (or other private secrets) from the log?
* [x] Is it rather better to upgrade to postgresql 11 and use the logical replication solution?

Once those questions are addressed we should:

* ~[ ] Setup client certificate based authentication and drop all secrets from database~
* [x] Enable `wal_level = replica` (https://www.postgresql.org/docs/9.6/continuous-archiving.html)
* [x] Create a base backup using "file-level" API to avoid x2 disk space usage during backup and upload it to s3 in `xz` format 
* [x] Define an archive_command which uploads the incremental update files to s3.
* [x] Document how third parties should sync it (https://www.postgresql.org/docs/9.6/continuous-archiving.html#BACKUP-PITR-RECOVERY & https://www.postgresql.org/docs/9.6/continuous-archiving.html#BACKUP-STANDALONE)
* [x] Validate and test that it works as expected
* ~[ ] Setup in airflow a task to run the backup on a monthly basis~

**Comment by hellais** [comment]
We discussed with @darkk a bit about the point:

>  If we go the WAL log route, how do we sanitise/remove user password hashes (or other private secrets) from the log?

We came to the conclusion that perhaps the best approach is to not have any secrets to begin with stored inside of the database and rely on other mechanisms not stored in the database for authentication.

This basically entails having tighter firewall rules for connecting to the database and enforcing [TLS client certificate based authentication](https://www.postgresql.org/docs/9.2/auth-methods.html#AUTH-CERT).

We could then maybe set the `PASSWORD` for the users as `NULL` to avoid mistaking the user password as a security mechanism or alternatively setting it to a string which starts with the `insecure_` prefix.

**Comment by darkk** [comment]
> Is it rather better to upgrade to postgresql 11 and use the logical replication solution?

My opinion on that matter is following:

- upgrade to PG 11 may be a good idea, it's also mentioned in #240, but I'd rather consider doing that as a separate step when the DR procedure is in place and tested :)
- using logical replication looks like wrong solution for this specific problem

Shipping WAL via S3 allows every consumer to use the S3 bucket as an independent source of the events to replicate, so the producer (publisher, writer, leader) does not have to keep old WAL files and the subscriber does not even need to know that publisher exists.

If I understand logical replication correctly, it needs direct link between publisher and subscriber, so the beauty of S3 storage will be gone in this case.

**Comment by darkk** [comment]
Given:
- 20...30 MB/s download speed for `autoclaved/jsonl.tar.lz4` files from [ooni-data](https://ooni-data.s3.amazonaws.com) bucket to GH AMS
- database filesystem snapshot ~ 590 GiB
- growth rate: 0.6 GiB/day
- very variable compression ratio of randomly chosen 1 GiB file from `${pgdir}/base`

| compressor   | ratio         | level for bw test | deflate  | inflate  |
| ------------ | ------------- | ----------------- | -------- | -------- |
| lz4 -{3..9}  | 3.10 ... 3.27 | lz -3             | 50 MB/s  | 850 MB/s |
| gzip -{1..9} | 2.85 ... 3.47 | gzip -6           | 22 MB/s  | 100 MB/s |
| xz -9        | 17.9          | xz -9             | 2.1 MB/s | 125 MB/s |

I believe, that some further experimentation with the data is needed to understand
- if `xz` is really worth the CPU spent
- if `gzip` saturates CPU during restore instead of network / disk
- if `lz4` is better than `gzip` in that aspect (decompressor costs ~nothing in terms of CPU)

BTW, compression ratio variance for WAL `pg_xlog` files is significantly lower:

| compressor  | ratio |
| ----------- | ----- |
| lz4 -{3..9} |  1.34 |
| gzip -{6,9} |  1.46 |
| xz -9       |  1.69 |

**Comment by hellais** [comment]
@darkk based on the numbers you provide for the compression of the base snapshot, I am leaning towards either going for a highly standard compressor (ex. gzip) or no compression at all.

If we put things in s3 and there is the need to do compression while the data is being retrieved, it can be enabled using the `content-encoding: gzip` option.

For the pg_xlog WAL files I think it's not worth to invest the effort to obtain such low compression ratios.

I am also noting in here what I think are the necessary steps to do this, with some questions I have as going through the docs:

* Set `wal_level = replica` (do we need to put into archive, according to the docs replica is good enough) and `archive_mode = on` in `/srv/pl-psql/postgresql.conf` (https://www.postgresql.org/docs/9.6/continuous-archiving.html)

* Define an `archive_command` which uploads the incremental update files to s3.

Question: do we want to upload them directly to s3 or stage them into a directory and upload them in batches? Are there issues with a long running `archive_command`?

* Create a base backup.
Question: do we want to use `pg_basebackup` or the low level API?

* Document how third parties should sync it. 
Question: should this happen using the "Recovering Using a Continuous Archive Backup" procedure: https://www.postgresql.org/docs/9.6/continuous-archiving.html#BACKUP-PITR-RECOVERY or "Standalone hot backups" https://www.postgresql.org/docs/9.6/continuous-archiving.html#BACKUP-STANDALONE?

**Comment by darkk** [comment]
> a highly standard compressor (ex. gzip) or no compression at all

Here are [wall-clock hours to download ~full snapshot](https://nbviewer.jupyter.org/gist/darkk/b351fd25e71b0be4a65555a8ff20caf5/pgzz.ipynb#Wall-clock-hours-to-download-from-S3).

We're discussing time-to-fetch differing by a factor of 17, ranging from 16.4 to 0.9 hours for HKG (~100 Mbit/s), ranging from 8.3h to 0.5h for AMS (~250 Mbit/s).

I hope, those numbers will make the decision a more informed one.

> do we want to upload them directly to s3 or stage them

I see no value in staging them given the added complexity. It may (unlikely) improve compression ratio, but it does not have significant effect. It'll require tracking of upload failures in the script itself.

> issues with a long running `archive_command`?

I'm unaware of any issues like that.

> do we want to use `pg_basebackup` or the low level API

Given the significant size of the database, I'd suggest to use "file-level" API to avoid x2 disk space usage during backup.

> should this happen using the "Recovering Using a Continuous Archive Backup" procedure or "Standalone hot backups"

IMHO, those are two sides of the same coin. I'd suggest to codify the procedure as a script that restores canonical PG version from S3 and enables periodic polling of S3 bucket for new WAL files.

**Comment by hellais** [comment]
> We're discussing time-to-fetch differing by a factor of 17, ranging from 16.4 to 0.9 hours for HKG (~100 Mbit/s), ranging from 8.3h to 0.5h for AMS (~250 Mbit/s).

Thanks for linking to that notebooks it contains very useful information.

From what I can tell it seems like gzip with level 5 (or even 4) is pretty much comparable to lz4, so I would tradeoff a bit of performance in exchange for making it easier for people to ingest the data (gzip is available by default on most systems).

Does this seem reasonable?

I also agree with all your other points. I am going to edit the top level issue to include what I think are the next steps for this.

**Comment by darkk** [comment]
Yep, I agree that `lz4` and `gzip` have no significant difference &mdash; both are limited by the network, so there is no significant gain in time-to-recover if `lz4` is used instead of `gzip` within these two (or slower) networks, just some minor power savings :-)

The option I suggest to consider is the `xz` (LZMA) at different levels. `xz` has outstanding compression ratio for the data files.  It's the compressor that is also available on many modern Linuxes, e.g. Debian stretch (current stable) uses it for some packaging duties. It's safe to say "Linux" as PosgreSQL data files are tied to (OS, CPU, PG) versions and are not guaranteed to work on another platform, as far as I know. [Logical replication is the way for cross-platform replication](https://www.postgresql.org/docs/10/logical-replication.html).

Do you think that CPU spent on backup is not worth x5 decrease of time-to-restore?

**Comment by hellais** [comment]
> Do you think that CPU spent on backup is not worth x5 decrease of time-to-restore?

I see the tradeoffs between the two approaches being (correct me if I am misreading the data):

A. gzip level4: It takes us 5.5 hours to compress the full base backup, a consumer of the DB takes 2.5 hours to download + 1.7h to decompress
B. xz level4: It takes us 47.4 hours to compress the full base backup, a consumer of the DB takes 0.6hours to download the data + 1.9h to decompress

I would say that both the archival (as you say done every second or third month) is done at the same interval as the restoring process for users synching the data.

I think if we invest the extra CPU time depends on how many consumers of the data we expect to have to make it worth our CPU cycles.

Ex. if there are only 2 consumers, we spend 47.4h to then give them a gain of only about 4h total. I think we only really start to break even on the CPU cycles once we get a number of consumers in the order of the dozens.

In any case I think that switching from `gzip` to `xz` is something that is probably cheap and easy to as the number of users of this sort of dataset grows.

Does this make sense?

That said I am also OK with going for `xz` from day 1 and spending those CPU hours.

**Comment by darkk** [comment]
The tradeoff is slightly different, you're not accounting for two factors: a) upload and compression are parallel processes so they're not summed but maximum of two is taken, b) compression task may be split across several CPU cores, so it's measured in CPU-hours, not wall-clock-hours. So these tradeoff options are closer to
A. gz:4 ~ 5.5h to compress and upload, 2.5h to download and decompress to AMS (not 4.2h as 2.5h + 1.7h suggests)
B. xz:2 ~ 5h to compress & upload from HKG, ~ 0.8h to download and decompress (assuming that we have 3 cores to spend on compression/decompression)

> I would say that both the archival ... is done at the same interval as the restoring process for users synching the data.

I don't understand why you say so. The user may have continuous restoration polling for new WAL segments from S3, so the willing consumer will always have "reasonably fresh" data if we call `pg_switch_xlog()` after bucket ingestion (or control the same via `archive_timeout`).

So the main `pg_base` usage is to bootstrap a new replica:
1. e.g. when new data consumer wants to have a local replica
1. e.g. when user asks for [Metadata as a service](https://ooni.torproject.org/post/mining-ooni-data/#metadata-as-a-service) instance
1. e.g. for disaster recovery after unfortunate `DELETE` being replicated

The first case probably does not benefit from speedy restore, the second and the third do. I'm unsure what's the correct multiplication factor to equate CPU-hour burned and Human-hour waiting during DR :)

Periodic full-dump is required to limit amount of WAL that has to be consumed to get "fresh" snapshot, the period of period full-dump is mostly determined by the limit we want to enforce and amount of daily WAL.

I also remember `aws s3 cp` having some pathological performance for uploading making it bottlenecked by CPU, that's also something worth understanding while implementing upload.

**Comment by hellais** [comment]
> So these tradeoff options are closer to

Ah yes I think you are right. Thanks for explaining this.

I think in light of this it makes most sense to go for `xz` with level 2 compression.

> I don't understand why you say so.

What I meant with that is that if I am using currently using the metadata as a service, I will be downloading the pg_base only once every time it is re-done by us as in the interim I will just be consuming the incremental WAL updates.

I do however agree that it's also pretty important to be able to do disaster recover based on this and to make the bootstrap time for new users more pleasant, so it's a good idea to keep that in mind.

**Comment by darkk** [comment]
> I will be downloading the pg_base only once every time it is re-done

I don't understand the reason to do that. WAL is a continuous stream of WAL files. As soon as the replica is bootstrapped and is _reasonably up-to-date_ there is no need to restore it from `base` dump.

I say "_reasonably up-to-date_" as WAL replay also takes resources besides network transfer. In a worst case WAL may be lots of random-writes and it may be cheaper to download huge batch of files then to apply WAL. We've seen alike pattern yesterday, when simple `UPDATE SET anomaly` for a 32GiB table did not converge for 48h, but corresponding table rebuild took less than 1h. The index over `input_no` took ~20m to build, so you can estimate it for other columns.

The UPDATE was touching ~7% of rows (17M out of 230M), but for various reasons PG did not plan UPDATE part in a sequental-scan manner. I _assume(!)_ that the main reason is the need to keep indexes in sync.

**Comment by darkk** [comment]
Okay, so here are the tables for `VACUUM FULL` to ensure that shared PG files are 100% clear after [the incident](https://ooni.torproject.org/post/ooni-probe-android-200-incident/):
- [ ] http_request, 182 GB
- [ ] http_control, 129 GB
- [ ] dns_a, 52 GB
- [ ] measurement, 33 GB
- [ ] http_verdict, 16 GB
- [ ] tcp, 12 GB
- [ ] report, 2001 MB
- [ ] residual, 1697 MB
- [ ] vanilla_tor, 166 MB
- [ ] http_request_fp, 98 MB
- [ ] input, 84 MB
- [ ] autoclaved, 40 MB
- [ ] ooexpl_bucket_msm_count, 31 MB
- [ ] originas, 28 MB
- [ ] domain, 10 MB
- [ ] ooexpl_daily_msm_count, 6296 kB
- [ ] ooexpl_netinfo, 1752 kB
- [ ] repeated_report, 184 kB
- [ ] software, 152 kB
- [ ] fingerprint, 48 kB
- [ ] ooexpl_website_msm, 16 kB
- [ ] badrow, 16 kB
- [ ] badblob, 16 kB
- [ ] report_files, 8192 bytes
- [ ] ng19verdict, 8192 bytes
- [ ] ng19msm, 8192 bytes
- [ ] badmeta, 8192 bytes

**Comment by darkk** [comment]
> BTW, compression ratio variance for WAL `pg_xlog` files is significantly lower

The possible reason is that I could use non-representative sample of WAL files to emit those statements as a series of 115 consecutive WAL files was used for the experiment, while 10 random 1GiB-sized base files were used.

E.g. minimal [`xz -2` compression ratio for WAL](https://nbviewer.jupyter.org/gist/darkk/b351fd25e71b0be4a65555a8ff20caf5/pgzz.ipynb#WAL-files) was 0.37 during the experiment, but randomly sampled WAL file from `http_request` VACUUMing gives 0.17(!). And 0.17 is more aligned with compression ration observed for those data files sample.

**Comment by darkk** [comment]
Per-bucket compression ratios show that the sample for the experiment was not super-representative:

| `xz -2`     | mean | min | 25% | 50% | 75% | max |
| ------- | ----:| ---:| ---:| ---:| ---:| ---:|
| sample | 0.08 | 0.02| 0.07| 0.09| 0.10| 0.16|
| `base` | 0.15 | 0.01| 0.07| 0.11| 0.15| 0.58|

The whole `base` comresses from 599 GiB to 89 GiB within 7h of 4-CPU run-time. The `xz` binary differs between the experiment and the actual container, but it's unclear if the difference between expected 15 cpu-hours and actual 28 cpu-hours is explained by that fact, by bad sampling or by some other mistake.

**Comment by darkk** [comment]
Time-to-restore was 80m for 4-CPU a EC2 instance in `us-east-1`. WAL for 1 bucket took ~5m more to process.

The way to bootstrap a new independent (without OONI-specific nuts and bolts) read-only replica on that EC2 instance was the following one:

```
$ : lsblk && mkfs at EBS && mount it to /mnt && mkdir /mnt/metadb
$ time ./metadb-restore DEST=/mnt/metadb fetch
$ wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | sudo apt-key add -
$ sudo add-apt-repository "deb http://apt.postgresql.org/pub/repos/apt/ $(lsb_release --short --codename)-pgdg main"
$ sudo apt-get install postgresql-9.6
$ sudo systemctl stop postgresql@9.6-main.service
$ sudo mv /var/lib/postgresql/9.6/main /var/lib/postgresql/9.6/main.dist
$ sudo chown -R postgres:postgres /mnt/metadb
$ sudo chmod 0700 /mnt/metadb
$ sudo ln -s /mnt/metadb /var/lib/postgresql/9.6/main
$ echo "hot_standby = 'on'" | sudo -u postgres dd of=/etc/postgresql/9.6/main/conf.d/hot_standby.conf
$ sudo systemctl start postgresql@9.6-main.service
$ sudo -u postgres psql -U postgres metadb -c 'SELECT MAX(bucket_date) FROM autoclaved'
```

**Comment by hellais** [comment]
@darkk is
> $ time ./metadb-restore DEST=/mnt/metadb fetch
This: https://github.com/ooni/sysadmin/blob/a7bce33e3d3a729388d9e74970030a91310b75fe/scripts/metadb_s3_tarx?

**Comment by darkk** [comment]
Yep. (I updated the comment).

**Comment by hellais** [comment]
Documentation on how to do this setup is written here: https://github.com/ooni/sysadmin/blob/metadb-share/docs/metadb-sharing.md

**Comment by darkk** [comment]
BTW, I've checked replication lag at EC2 instance today. `meta_db` task was done by 05:03, but the final WAL piece (containing `COMMIT`) was not shipped & fetched till I've triggered WAL switch manually at ~10:00. There are several ways to handle that:
- tune `archive_timeout` + `checkpoint_timeout`, but [_it is unwise to use a very short archive_timeout — it will bloat your archive storage_](https://www.postgresql.org/docs/9.6/runtime-config-wal.html)
- explicitly flush WAL after pipeline cycle with `pg_switch_xlog()`, but it needs superuser
- ignore that and live with 24h replication lag for WAL-shipment-over-S3 case

**Comment by darkk** [comment]
@pellaeon brought the case that he is unable to _restart(!)_ postgresql replica having "up-to-date" WAL, suggesting that we may be affected by one of [Hot Standby Caveats](https://www.postgresql.org/docs/9.6/hot-standby.html#HOT-STANDBY-CAVEATS). And the following one kinda fits:

> Full knowledge of running transactions is required before snapshots can be taken. Transactions that use large numbers of subtransactions (currently greater than 64) will delay the start of read only connections until the completion of the longest running write transaction.

It's unclear to me what "subtransactions" is, but I suspect that _"savepoint"_ is probably a synonym for "subtransaction". It's also unclear to me if a released savepoint counts as a "running transaction" or not. `pg_xlogdump` shows that `Transaction` ingesting a bucket has 283 `subxacts`, so, it sounds like a possible cause to me.

Also, I was able to reproduce inability to restart a slave when the restart was reaching the point `consistent recovery state reached` but not the point `database system is ready to accept read only connections` and `pg_switch_xlog()` "fixed" my instance but not @pellaeon's.

IMHO, the case shows that _"ignore that and live with 24h replication lag"_ is not an option, so I'm adding explicit `pg_switch_xlog()` as a pipeline step after `meta_pg` as it's an easy and understandable action.

**Comment by hellais** [comment]
From the perspective of this being usable by third parties I would consider this done. As mentioned by @darkk I am going to be adding a periodic operation to create a new DB dump every month.

The original plan was to do it as part of the airflow scheduler, but I am wondering if perhaps a quicker and easier option is to do it with a simple cronjob running on the DB host.

In any case no further action from @darkk is needed on this item.

**Comment by hellais** [comment]
I migrated this task into another issue: https://github.com/ooni/sysadmin/issues/309, so we can close this.
